### PR TITLE
mewo

### PR DESCRIPTION
**Pull request by Hiramiya**
This adds two new commands to Sakuya; `mewo` and `mewoset`, and an option to enable/disable the feature per-guild (via `enable`/`disable` commands, requires Manage Server)

`mewo [user]` has Sakuya respond with either the custom mewo response saved against your member ID, the one of the member in an `@mention`, or a random one if one is not set for the member.
`mewoset [user] <mewo>` sets the custom response either for yourself or the `@mention` member. This can only be set by those with the Manage Nicknames permission.

### CODE REVIEW

File: sakuya/mewo.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+from discord.ext import commands
+from discord.ext.commands import Bot
+
+from .db import db, Guild as _Guild, Member as _Member
+
+
+random_mewo = [
+    "ME.. oh you're not mcmower.",
+    "MEWO",
+    "You didn't really expect me to say mewo did you"
```
Comment (mcpower) [inline sakuya/mewo.py:15]: style nit: add leading comma here for cleaner diffs when adding more items to the list

File: sakuya/mewo.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
+from dataclasses import dataclass
```
Comment (mcpower) [inline sakuya/mewo.py:1]: style nit for the entire file (why doesn't GitHub allow file-level comments...): I believe this project (mostly) uses single quotes, so change all double-quotes to single-quotes in this file

File: sakuya/mewo.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    @commands.command()
+    @commands.has_permissions(manage_nicknames=True)
+    async def mewoset(self, ctx, *, mewo_text):
+        state = self.guilds.get(ctx.guild)
+        if state:
+
+            if len(ctx.message.mentions) == 1:
+                memberid = ctx.message.mentions[0].id
+                print(mewo_text)
+                mewo_text = str(mewo_text.split(" ",1)[1])
```
Comment (mcpower) [inline sakuya/mewo.py]: Is it guaranteed that there's a space in `mewo_text`? If not, the split here will return a list with one item, causing an IndexError upon indexing.

Additionally, consider using the default split behaviour by omitting the separator argument - it splits on runs of whitespace and discards empty strings:

```python
>>> '  MEWO MCMOWER'.split(' ', 1)
['', ' MEWO MCMOWER']
>>> '  MEWO MCMOWER'.split(maxsplit=1)
['MEWO', 'MCMOWER']
```

File: sakuya/mewo.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+            memberid = ctx.message.mentions[0].id if len(ctx.message.mentions) == 1 else ctx.author.id
+
+            member = db.query(_Member).get((memberid, ctx.guild.id)) or _Member(user_id=memberid,
+                                                                                guild_id=ctx.guild.id)
+
+            if member.mewo_text:
+                msg = member.mewo_text
+            else:
+                msg = random.choice(random_mewo)
```
Comment (mcpower) [inline sakuya/mewo.py]: nit: `or` would work here. However, it's arguably less readable...

@Trif4 What do you think?

File: sakuya/mewo.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                print(mewo_text)
+                mewo_text = str(mewo_text.split(" ",1)[1])
+                pronoun = "their"
+            else:
+                memberid = ctx.author.id
+                pronoun = "your"
+
+            member = db.query(_Member).get((memberid, ctx.guild.id)) or _Member(user_id=memberid,
+                                                                                guild_id=ctx.guild.id)
+            previous_mewo = member.mewo_text
```
Comment (mcpower) [inline sakuya/mewo.py:81]: nit: You could inline this variable as `member.mewo_text` isn't set before `previous_mewo` is read from, but having the explicit variable name is more readable...

@Trif4 What do you think?

File: sakuya/mewo.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                else:
+                    msg = f"I have set {pronoun} mewo response."
+
+                member.mewo_text = mewo_text
+                db.add(member)
+                db.commit()
+
+                await ctx.send(msg)
+
+            except (AssertionError) as e:
```
Comment (mcpower) [inline sakuya/mewo.py]: style nit: parentheses here are redundant

File: sakuya/mewo.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                await ctx.send("I'm terribly sorry, but I'm unable to do that at the moment. Please try again later.")
+                print(e)
+
+    async def enable(self, ctx):
+        self.guilds[ctx.guild] = GuildState(guild=ctx.guild, mewo_enabled=True)
+        g = db.query(_Guild).get(ctx.guild.id) or _Guild(id=ctx.guild.id)
+        g.mewo_enabled = True
+        db.add(g)
+        db.commit()
+        await ctx.send(f'Mewo responses enabled.')
```
Comment (mcpower) [inline sakuya/mewo.py]: style nit: this f-string can be a normal string

**Review by mcpower — COMMENTED** [review]
MEWO HEWOMEOWNYA

Mostly minor style nitpicks - I'm used to "must keep repo perfectly styled at all times" so you can ignore them if Trif doesn't care 😛

File: sakuya/mewo.py (hunk 1 above)
Comment (Hiramiya) [reply sakuya/mewo.py:15]: Other files in the repo exclude a trailing comma, matching their style

File: sakuya/mewo.py (hunk 2 above)
Comment (Hiramiya) [reply sakuya/mewo.py:1]: Quotes follow the same-ish pattern from other files in the project; where fixed-strings are generally single-quotes and formated strings/client responses are double

File: sakuya/mewo.py (hunk 3 above)
Comment (Hiramiya) [reply sakuya/mewo.py]: Completely forgot about named arguments in built-in functions; there _should_ always be a whitespace separator, and is usually a space in the discord client, but no reason not to change this

File: sakuya/mewo.py (hunk 4 above)
Comment (Hiramiya) [reply sakuya/mewo.py]: Ah, this is leftover from when this was a slightly longer and more complicated `if`. Updated for now, can always be changed back if the slightly higher readability is wanted

File: sakuya/mewo.py (hunk 5 above)
Comment (Hiramiya) [reply sakuya/mewo.py:81]: This one is also due to changes in code arrangement between testing; previously the member lookup and response were in a different order that neccessitated the use of an additional var, but that caused a different issue (hence the re-order). I do like this one for the readability aspect, but will leave @Trif4 to decide

File: sakuya/mewo.py (hunk 7 above)
Comment (Hiramiya) [reply sakuya/mewo.py]: Caught this one before and hadn't committed, oops!

File: sakuya/mewo.py (hunk 6 above)
Comment (Hiramiya) [reply sakuya/mewo.py]: Fixed!

File: sakuya/mewo.py (hunk 5 above)
Comment (Trif4) [reply sakuya/mewo.py:81]: Personally I'd inline it; you can still preserve readability by just adding a comment like "User already has a mewo".

**Review by Trif4 — CHANGES_REQUESTED** [review]
I like this & appreciate the PR, but I think the feature needs to be changed a little to fit the spirit of mewo better.

I think it should trigger when you say "mewo [bot prefix]", to which the bot responds "MEWO [user mewo]". The user's "mewo" should be a mewoified nickname, set with `mewoset` as already implemented. If no custom mewo is stored for the user, the bot should attempt to mewoify their current server nickname on its own.

The toughest part of this is that the commands system doesn't support invocation without a command prefix. It's possible to just read every message and check for "mewo [bot prefix]", but it'd probably require changes to the project structure. Maybe some system for cogs to register prefix-less commands with a global message-reading hook? (Take care that normal command parsing still runs if you use such a hook!)

**Comment by Trif4** [comment]
Closing as stale, feel free to reopen if you decide to pick this up again.

**Pull request closed without merging**